src/coreclr/vm/threadstatics.cpp

```diff
@@ -405,6 +405,13 @@ void FreeLoaderAllocatorHandlesForTLSData(Thread *pThread)
                 }
             }
         }
+
+        pThread->cLoaderHandles = -1; // Sentinel value indicating that there are no LoaderHandles and the thread is permanently dead.
+        if (pThread->pLoaderHandles != NULL)
```

Comment:
Nit: This should be always true since `cLoaderHandles > 0` here. Also, the `delete` can be called outside the lock.